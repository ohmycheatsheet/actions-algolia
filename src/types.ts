export type Issue = {
  id: string
  number: string
  title: string
  body: string
  createdAt: string
  updatedAt: string
  url: string
  state: 'open' | 'state'
  labels: Label[]
}

export type Label = {
  id: string
  color: string
  description: string
  name: string
  default: boolean
  createdAt: string
  updatedAt: string
}
