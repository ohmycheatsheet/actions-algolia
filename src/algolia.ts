import algoliasearch from 'algoliasearch'

const issueClient = algoliasearch(process.env.ALGOLIA_APPID!, process.env.ALGOLIA_APP_KEY!)

const labelClient = algoliasearch(process.env.ALGOLIA_APPID!, process.env.ALGOLIA_APP_KEY!)

export const algolia = {
  async ensureInit() {
    labelClient.initIndex('actions_cheatsheet_labels')
    issueClient.initIndex('actions_cheatsheet_issues')
  },
}
