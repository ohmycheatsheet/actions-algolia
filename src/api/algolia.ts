import algoliasearch, { SearchIndex } from 'algoliasearch'
import { Issue, Label } from '../types'

const issueClient = algoliasearch(process.env.ALGOLIA_APPID!, process.env.ALGOLIA_APP_KEY!)
const labelClient = algoliasearch(process.env.ALGOLIA_APPID!, process.env.ALGOLIA_APP_KEY!)

let issueIndex: SearchIndex
let labelIndex: SearchIndex

export const algolia = {
  async ensureInit() {
    labelIndex = labelClient.initIndex('actions_cheatsheet_labels')
    issueIndex = issueClient.initIndex('actions_cheatsheet_issues')
    issueIndex.setSettings({
      searchableAttributes: ['title', 'description', 'body'],
      attributesForFaceting: ['state', 'filterOnly(labels.name)', 'filterOnly(labels.id)'],
      customRanking: ['desc(updatedAt_timestamp)'],
    })
    labelIndex.setSettings({
      searchableAttributes: ['name', 'description'],
    })
  },
  async uploadCheatsheets(cheatsheets: Issue[]) {
    await this.ensureInit()
    issueIndex.saveObjects(cheatsheets, {
      autoGenerateObjectIDIfNotExist: true,
    })
  },
  async uploadTags(tags: Label[]) {
    await this.ensureInit()
    labelIndex.saveObjects(tags, {
      autoGenerateObjectIDIfNotExist: true,
    })
  },
}
