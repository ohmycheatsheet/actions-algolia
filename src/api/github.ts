import { graphql } from '@octokit/graphql'
import { Issue, Label } from '../types'
import dayjs from 'dayjs'

export const gql = String.raw

const graphqlWithAuth = graphql.defaults({
  headers: {
    authorization: `token ${process.env.GITHUB_TOKEN}`,
  },
})

export const github = {
  issueCount: async (owner: string, name: string): Promise<number> => {
    const response: any = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            issues {
              totalCount
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response.repository.issues.totalCount
  },
  labelCount: async (owner: string, name: string): Promise<number> => {
    const response: any = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            labels {
              totalCount
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response.repository.labels.totalCount
  },
  labels: async (
    owner: string,
    name: string,
    after?: string,
  ): Promise<{
    labels: Label[]
    pageInfo: {
      endCursor: string
      hasNextPage: boolean
    }
  }> => {
    const response: any = await graphqlWithAuth(
      gql`
        query Labels($owner: String!, $name: String!, $after: String) {
          repository(owner: $owner, name: $name) {
            labels(first: 10, after: $after) {
              edges {
                node {
                  id
                  description
                  name
                  color
                }
              }
              pageInfo {
                endCursor
                hasNextPage
              }
            }
          }
        }
      `,
      {
        owner,
        name,
        after,
      },
    )
    console.log(response)
    const labels = response.repository.labels.edges.map((item: { node: any }) => ({
      ...item.node,
      // date -> timestamp
      updatedAt_timestamp: dayjs(item.node.updatedAt).unix(),
      createdAt_timestamp: dayjs(item.node.createdAt).unix(),
    }))
    const pageInfo = response.repository.labels.pageInfo
    return {
      labels,
      pageInfo,
    }
  },
  issues: async (
    owner: string,
    name: string,
    after?: string,
  ): Promise<{
    issues: Issue[]
    pageInfo: {
      endCursor: string
      hasNextPage: boolean
    }
  }> => {
    const response: any = await graphqlWithAuth(
      gql`
        query Issues($owner: String!, $name: String!, $after: String) {
          repository(owner: $owner, name: $name) {
            issues(first: 10, after: $after, states: OPEN) {
              edges {
                node {
                  id
                  number
                  title
                  body
                  createdAt
                  updatedAt
                  state
                  labels(last: 10) {
                    edges {
                      node {
                        id
                        color
                        description
                        name
                        createdAt
                        updatedAt
                      }
                    }
                  }
                }
              }
              pageInfo {
                endCursor
                hasNextPage
              }
            }
          }
        }
      `,
      {
        owner,
        name,
        after,
      },
    )
    console.log(response.repository.issues.edges)
    const issues = response.repository.issues.edges.map((item: { node: any }) => ({
      // unzip issue connection
      ...item.node,
      // unzip labels
      labels: item.node.labels.edges.map((label: { node: any }) => label.node),
      // date -> timestamp
      updatedAt_timestamp: dayjs(item.node.updatedAt).unix(),
      createdAt_timestamp: dayjs(item.node.createdAt).unix(),
    }))
    const pageInfo = response.repository.issues.pageInfo
    return {
      issues,
      pageInfo,
    }
  },
}
