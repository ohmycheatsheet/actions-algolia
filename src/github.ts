import { graphql } from '@octokit/graphql'

const graphqlWithAuth = graphql.defaults({
  headers: {
    authorization: `token ${process.env.GITHUB_TOKEN}`,
  },
})

export const api = {
  issueCount: async (owner: string, name: string) => {
    const response = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            issues {
              totalCount
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response
  },
  labelCount: async (owner: string, name: string) => {
    const response = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            labels {
              totalCount
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response
  },
  labels: async (owner: string, name: string) => {
    const response = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            labels(first: 10) {
              edges {
                node {
                  id,
                  description,
                  name,
                  color
                }
              }
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response
  },
  issues: async (owner: string, name: string) => {
    const response = await graphqlWithAuth(
      `
        query Issues($owner: String!, $name: String!) {
          repository(owner: $owner, name: $name) {
            issues(last: 3, states: OPEN) {
              edges {
                node {
                  id,
                  number
                  title,
                  body,
                  createdAt,
                  updatedAt,
                  state,
                  labels(last: 10) {
                    edges {
                      node {
                        id,
                        color,
                        description,
                        name,
                        createdAt,
                        updatedAt
                      }
                    }
                  }
                }
              }
            }
          }
        }
      `,
      {
        owner,
        name,
      },
    )
    console.log(response)
    return response
  },
}
