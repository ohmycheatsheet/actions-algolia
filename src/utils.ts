import { Issue, Label } from './types'
import dayjs from 'dayjs'
import omit from 'lodash.omit'

type LabelNode = { node: Label }
type IssueNode = { node: Issue }

export const normalize = {
  label(label: LabelNode) {
    return {
      ...label.node,
      objectID: label.node.id,
      updatedAt_timestamp: dayjs(label.node.updatedAt).unix(),
      createdAt_timestamp: dayjs(label.node.createdAt).unix(),
    }
  },
  labels(labels: LabelNode[]) {
    return labels.map(this.label)
  },
  issue(issue: Issue) {
    const labels = (issue.labels as any).edges.map((label: any) =>
      omit(this.label(label), ['objectID', 'updatedAt_timestamp', 'createdAt_timestamp']),
    )
    return {
      ...issue,
      objectID: issue.id,
      labels,
      // date -> timestamp
      updatedAt_timestamp: dayjs(issue.updatedAt).unix(),
      createdAt_timestamp: dayjs(issue.createdAt).unix(),
    }
  },
  issues(issues: IssueNode[]) {
    return issues.map(issue => normalize.issue(issue.node))
  },
}
