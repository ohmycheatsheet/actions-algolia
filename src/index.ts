import * as core from '@actions/core'
import { github } from './github'
import { algolia } from './algolia'

const syncIssues = async (owner: string, name: string) => {
  let { issues, pageInfo } = await github.issues(owner, name)
  let after
  for (; ; after = pageInfo.endCursor) {
    ;({ issues, pageInfo } = await github.issues(owner, name, after))
    const cheatsheets = issues.map(item => {
      return {
        ...item,
        objectID: item.id,
      }
    })
    await algolia.uploadCheatsheets(cheatsheets)
    if (!pageInfo.hasNextPage) {
      break
    }
  }
}

const syncLabels = async (owner: string, name: string) => {
  let { labels, pageInfo } = await github.labels(owner, name)
  let after
  for (; ; after = pageInfo.endCursor) {
    ;({ labels, pageInfo } = await github.labels(owner, name, after))
    const tags = labels.map(item => {
      return {
        ...item,
        objectID: item.id,
      }
    })
    await algolia.uploadTags(tags)
    if (!pageInfo.hasNextPage) {
      break
    }
  }
}

// most @actions toolkit packages have async methods
async function run() {
  try {
    const repo = process.env.GITHUB_REPOSITORY
    const [owner, name] = repo?.split('/') || []
    await syncIssues(owner, name)
    await syncLabels(owner, name)
  } catch (error) {
    console.log(error)
    core.setFailed((error as any).message)
  }
}

run()
