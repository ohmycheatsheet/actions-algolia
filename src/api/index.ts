import { algolia } from './algolia'
import { github } from './github'
import { normalize } from '../utils'

const syncIssues = async (owner: string, name: string) => {
  let { issues, pageInfo } = await github.issues(owner, name)
  let after
  for (; ; after = pageInfo.endCursor) {
    ;({ issues, pageInfo } = await github.issues(owner, name, after))
    await algolia.uploadCheatsheets(issues)
    if (!pageInfo.hasNextPage) {
      break
    }
  }
}

const syncLabels = async (owner: string, name: string) => {
  let { labels, pageInfo } = await github.labels(owner, name)
  let after
  for (; ; after = pageInfo.endCursor) {
    ;({ labels, pageInfo } = await github.labels(owner, name, after))
    await algolia.uploadTags(labels)
    if (!pageInfo.hasNextPage) {
      break
    }
  }
}

export const api = {
  github,
  algolia,
  schedule: async (owner: string, name: string) => {
    await syncIssues(owner, name)
    await syncLabels(owner, name)
  },
  issue: async (owner: string, name: string, number: number) => {
    const issue = await api.github.issue(owner, name, number)
    const tags = normalize.labels((issue.labels as any).edges)
    const cheatsheets = [normalize.issue(issue)]
    await algolia.uploadCheatsheets(cheatsheets)
    await algolia.uploadTags(tags)
    return cheatsheets
  },
}
