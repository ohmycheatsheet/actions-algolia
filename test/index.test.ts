import { api } from '../src/api'

describe('test', () => {
  it('fetch repo issues should work', async () => {
    const res = await api.github.issues('JiangWeixian', 'use-rematch')
    expect(res.issues).toMatchSnapshot()
    expect(res.pageInfo).toBeDefined()
  })

  it('fetch repo labels should work', async () => {
    const res = await api.github.labels('JiangWeixian', 'use-rematch')
    expect(res.labels).toMatchSnapshot()
    expect(res.pageInfo).toBeDefined()
  })

  it('fetch repo issue by issue number should work', async () => {
    const res = await api.github.issue('ohmycheatsheet', 'actions-algolia', 6)
    expect(res.number).toBe(6)
  })

  it('sync single issue to algolia should work', async () => {
    await api.algolia.ensureInit('actions-algolia')
    const issue = await api.issue('ohmycheatsheet', 'actions-algolia', 6)
    expect(issue).toMatchSnapshot()
  })
})
