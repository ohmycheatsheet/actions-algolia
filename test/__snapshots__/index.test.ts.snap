// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test fetch repo issues should work 1`] = `
Array [
  Object {
    "body": "**will do**

- [x] - 暴露出来的\`api\`需要规范统一
- [ ] - 提供类似\`usepage or useloadmore\`这些模型",
    "createdAt": "2020-08-02T13:56:49Z",
    "createdAt_timestamp": 1596376609,
    "id": "MDU6SXNzdWU2NzE2MjEwMDg=",
    "labels": Array [
      Object {
        "color": "0052cc",
        "createdAt": "2021-09-27T01:19:01Z",
        "description": "feature in mind",
        "id": "LA_kwDODtXGfs7KMCSr",
        "name": "rfc",
        "updatedAt": "2021-09-27T01:19:01Z",
      },
    ],
    "number": 10,
    "objectID": "MDU6SXNzdWU2NzE2MjEwMDg=",
    "state": "OPEN",
    "title": "rfc",
    "updatedAt": "2021-09-27T01:20:14Z",
    "updatedAt_timestamp": 1632705614,
  },
  Object {
    "body": "builtin partial update reducers like

\`\`\`ts
const model = createModel({
  state: {
    position
  } as State,
})

const { state, dispatch } = useRematch(model)
\`\`\`

this model will auto contain \`dispatch.update\` to partial update state, such as \`position\`",
    "createdAt": "2021-09-27T01:18:30Z",
    "createdAt_timestamp": 1632705510,
    "id": "I_kwDODtXGfs48DnMT",
    "labels": Array [
      Object {
        "color": "0052cc",
        "createdAt": "2021-09-27T01:19:01Z",
        "description": "feature in mind",
        "id": "LA_kwDODtXGfs7KMCSr",
        "name": "rfc",
        "updatedAt": "2021-09-27T01:19:01Z",
      },
    ],
    "number": 45,
    "objectID": "I_kwDODtXGfs48DnMT",
    "state": "OPEN",
    "title": "[rfc] auto builtin useful reducers",
    "updatedAt": "2021-09-27T01:20:58Z",
    "updatedAt_timestamp": 1632705658,
  },
]
`;

exports[`test fetch repo labels should work 1`] = `
Array [
  Object {
    "color": "d73a4a",
    "createdAt_timestamp": 1633708274,
    "description": "Something isn't working",
    "id": "MDU6TGFiZWwxOTIzOTEyNjI1",
    "name": "bug",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjI1",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "0075ca",
    "createdAt_timestamp": 1633708274,
    "description": "Improvements or additions to documentation",
    "id": "MDU6TGFiZWwxOTIzOTEyNjI4",
    "name": "documentation",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjI4",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "cfd3d7",
    "createdAt_timestamp": 1633708274,
    "description": "This issue or pull request already exists",
    "id": "MDU6TGFiZWwxOTIzOTEyNjMy",
    "name": "duplicate",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjMy",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "a2eeef",
    "createdAt_timestamp": 1633708274,
    "description": "New feature or request",
    "id": "MDU6TGFiZWwxOTIzOTEyNjM2",
    "name": "enhancement",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjM2",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "008672",
    "createdAt_timestamp": 1633708274,
    "description": "Extra attention is needed",
    "id": "MDU6TGFiZWwxOTIzOTEyNjM4",
    "name": "help wanted",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjM4",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "7057ff",
    "createdAt_timestamp": 1633708274,
    "description": "Good for newcomers",
    "id": "MDU6TGFiZWwxOTIzOTEyNjQx",
    "name": "good first issue",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjQx",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "e4e669",
    "createdAt_timestamp": 1633708274,
    "description": "This doesn't seem right",
    "id": "MDU6TGFiZWwxOTIzOTEyNjQz",
    "name": "invalid",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjQz",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "d876e3",
    "createdAt_timestamp": 1633708274,
    "description": "Further information is requested",
    "id": "MDU6TGFiZWwxOTIzOTEyNjQ1",
    "name": "question",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjQ1",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "ffffff",
    "createdAt_timestamp": 1633708274,
    "description": "This will not be worked on",
    "id": "MDU6TGFiZWwxOTIzOTEyNjQ2",
    "name": "wontfix",
    "objectID": "MDU6TGFiZWwxOTIzOTEyNjQ2",
    "updatedAt_timestamp": 1633708274,
  },
  Object {
    "color": "0366d6",
    "createdAt_timestamp": 1633708274,
    "description": "Pull requests that update a dependency file",
    "id": "MDU6TGFiZWwxOTM3ODMxNTQ0",
    "name": "dependencies",
    "objectID": "MDU6TGFiZWwxOTM3ODMxNTQ0",
    "updatedAt_timestamp": 1633708274,
  },
]
`;

exports[`test sync single issue to algolia should work 1`] = `
Array [
  Object {
    "body": "how actions aloglia work",
    "createdAt": "2021-09-18T01:12:44Z",
    "createdAt_timestamp": 1631927564,
    "id": "I_kwDOGC3jsM47l_0t",
    "labels": Array [
      Object {
        "color": "0075ca",
        "createdAt": "2021-09-12T14:14:11Z",
        "description": "Improvements or additions to documentation",
        "id": "MDU6TGFiZWwzMzQ2NTIxNTQ3",
        "name": "documentation",
        "updatedAt": "2021-09-12T14:14:11Z",
      },
      Object {
        "color": "7057ff",
        "createdAt": "2021-09-12T14:14:11Z",
        "description": "Good for newcomers",
        "id": "MDU6TGFiZWwzMzQ2NTIxNTU3",
        "name": "good first issue",
        "updatedAt": "2021-09-12T14:14:11Z",
      },
    ],
    "number": 6,
    "objectID": "I_kwDOGC3jsM47l_0t",
    "state": "CLOSED",
    "title": "Test issue",
    "updatedAt": "2021-09-30T01:53:36Z",
    "updatedAt_timestamp": 1632966816,
  },
]
`;
